onnxruntime/core/optimizer/matmul_transpose_fusion.cc

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/matmul_transpose_fusion.h"
+#include "core/graph/graph_utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+static std::pair<bool, Node*> IsInputTranspose(const Graph& graph, NodeArg& node_arg) {
+  auto trans_node = graph.GetProducerNode(node_arg.Name());
+  if (trans_node == nullptr || trans_node->OpType() != "Transpose") {
+    return std::make_pair(false, nullptr);
+  }
+
+  auto perms = RetrieveValues<int64_t>(trans_node->GetAttributes().at("perm"));
+  int64_t rank = perms.size();
+  if (rank < 2) {
+    return std::make_pair(false, nullptr);
+  }
+
+  bool is_trans_on_last_two_dims = true;
+  for (int64_t i = 0; i < rank - 2; i++) {
+    if ((int64_t)perms[i] != i) {
+      is_trans_on_last_two_dims = false;
+      break;
+    }
+  }
+
+  if (is_trans_on_last_two_dims) {
+    is_trans_on_last_two_dims = (int64_t)perms[rank - 2] == rank - 1 && (int64_t)perms[rank - 1] == rank - 2;
+  }
+
+  if (!is_trans_on_last_two_dims) {
+    return std::make_pair(false, nullptr);
+  }
+
+  return std::make_pair(true, const_cast<Node*>(trans_node));
+}
+
+static size_t UpdateConsumerCount(Graph& graph, NodeArg* target, std::unordered_map<NodeArg*, size_t>& count_map) {
+  const auto& node_consumers = graph.GetConsumerNodes(target->Name());
+  ORT_ENFORCE(!node_consumers.empty());
```

Comment:
Does this take into account if the output of the node potentially being removed provides a graph output like graph_utils::CanRemoveNode does?

Reply:
tested with the util and bert toy training test is broken. will keep the code the same for now
Investigate in future as needed. 
@pengwa / @souptc - FYI

---
In reply to: [397603399](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397603399) [](ancestors = 397603399)

Reply:
Sorry - don't quite understand the response. If you aren't checking if the node provides a graph output you may remove a node that you shouldn't and break the graph. How would adding that check cause something to break?

---
In reply to: [399855763](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r399855763) [](ancestors = 399855763,397603399)

Reply:
added the check

---
In reply to: [400542400](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r400542400) [](ancestors = 400542400,399855763,397603399)